File: Soe.Angular/TypeScript/Common/Models/RetroactivePayroll.ts
import { IRetroactivePayrollDTO, IRetroactivePayrollAccountDTO, IRetroactivePayrollEmployeeDTO, IAccountDimDTO, IRetroactivePayrollOutcomeDTO, IAttestPayrollTransactionDTO } from "../../Scripts/TypeLite.Net4";
import { Guid } from "../../Util/StringUtility";
import { SoeEntityState, TermGroup_SoeRetroactivePayrollStatus, TermGroup_RetroactivePayrollAccountType, TermGroup_SoeRetroactivePayrollEmployeeStatus, TermGroup_SoeRetroactivePayrollOutcomeErrorCode, TermGroup_PayrollResultType } from "../../Util/CommonEnumerations";
import { AccountDTO } from "./AccountDTO";
import { AccountDimDTO } from "./AccountDimDTO";
import { CalendarUtility } from "../../Util/CalendarUtility";

export class RetroactivePayrollDTO implements IRetroactivePayrollDTO {
    actorCompanyId: number;
    created: Date;
    createdBy: string;
    dateFrom: Date;
    dateTo: Date;
    modified: Date;
    modifiedBy: string;
    name: string;
    note: string;
    nrOfEmployees: number;
    retroactivePayrollAccounts: RetroactivePayrollAccountDTO[];
    retroactivePayrollEmployees: RetroactivePayrollEmployeeDTO[];
    retroactivePayrollId: number;
    state: SoeEntityState;
    status: TermGroup_SoeRetroactivePayrollStatus;
    statusName: string;
    timePeriodHeadId: number;
    timePeriodHeadName: string;
    timePeriodId: number;
    timePeriodName: string;
    timePeriodPaymentDate: Date;
    totalAmount: number;

    public fixDates() {
        this.dateFrom = CalendarUtility.convertToDate(this.dateFrom);
        this.dateTo = CalendarUtility.convertToDate(this.dateTo);
        this.timePeriodPaymentDate = CalendarUtility.convertToDate(this.timePeriodPaymentDate);
    }
}

export class RetroactivePayrollAccountDTO implements IRetroactivePayrollAccountDTO {
    accountDim: AccountDimDTO;
    accountDimId: number;
    accountId: number;
    created: Date;
    createdBy: string;
    modified: Date;
    modifiedBy: string;
    retroactivePayrollAccountId: number;
    retroactivePayrollId: number;
    state: SoeEntityState;
    type: TermGroup_RetroactivePayrollAccountType;

    //extensions
    guid: Guid;
    selectedAccount: AccountDTO;

    constructor() {
    }
}

export class RetroactivePayrollEmployeeDTO implements IRetroactivePayrollEmployeeDTO {
    actorCompanyId: number;
    categoryIds: number[];
    created: Date;
    createdBy: string;
    employeeId: number;
    employeeName: string;
    employeeNr: string;
    hasOutcomes: boolean;
    hasTransactions: boolean;
    modified: Date;
    modifiedBy: string;
    note: string;
    payrollGroupId: number;
    retroactivePayrollEmployeeId: number;
    retroactivePayrollId: number;
    retroactivePayrollOutcomes: IRetroactivePayrollOutcomeDTO[];
    state: SoeEntityState;
    status: TermGroup_SoeRetroactivePayrollEmployeeStatus;
    statusName: string;
    totalAmount: number;

    //extensions
    selected: boolean;
    moved: boolean;
    guid: Guid;

    constructor() {
    }
}

export class RetroactivePayrollOutcomeDTO implements IRetroactivePayrollOutcomeDTO {
    actorCompanyId: number;
    amount: number;
    created: Date;
    createdBy: string;
    employeeId: number;
    errorCode: TermGroup_SoeRetroactivePayrollOutcomeErrorCode;
    errorCodeText: string;
    isQuantity: boolean;
    isReadOnly: boolean;
    isRetroCalculated: boolean;
    isReversed: boolean;
    isSpecifiedUnitPrice: boolean;
    modified: Date;
    modifiedBy: string;
    productId: number;
    payrollProductName: string;
    payrollProductNumber: string;
    payrollProductNumberSort: string;
    payrollProductString: string;
    quantity: number;
    quantityString: string;
    resultType: TermGroup_PayrollResultType;
    retroactivePayrolIEmployeeId: number;
    retroactivePayrollOutcomeId: number;
    retroUnitPrice: number;
    specifiedUnitPrice: number;
    state: SoeEntityState;
    transactionUnitPrice: number;

    //extensions        
    retroDiffFormatted: string;
    amountFormatted: string;
    hasTransactions: boolean;
    transactions: IAttestPayrollTransactionDTO[];
    constructor() {
    }
}

File: Soe.Data/Models/RetroactivePayroll.cs
using SoftOne.Soe.Common.DTO;
using SoftOne.Soe.Common.Interfaces.Common;
using SoftOne.Soe.Common.Util;
using SoftOne.Soe.Data.Util;
using System;
using System.Collections.Generic;
using System.Linq;

namespace SoftOne.Soe.Data
{
    public partial class RetroactivePayroll : ICreatedModified, IState
    {
        public string StatusName { get; set; }
        public int NrOfEmployees { get; set; }
    }

    public partial class RetroactivePayrollEmployee : ICreatedModified, IState
    {

    }

    public partial class RetroactivePayrollOutcome : ICreatedModified, IState
    {
        public bool CalculatedHasTransactions { get; set; }
    }

    public static partial class EntityExtensions
    {
        #region RetroactivePayroll

        public static RetroactivePayrollDTO ToDTO(this RetroactivePayroll e)
        {
            if (e == null)
                return null;

            RetroactivePayrollDTO dto = new RetroactivePayrollDTO()
            {
                RetroactivePayrollId = e.RetroactivePayrollId,
                ActorCompanyId = e.ActorCompanyId,
                TimePeriodId = e.TimePeriodId,
                DateFrom = e.DateFrom,
                DateTo = e.DateTo,
                Name = e.Name,
                Status = (TermGroup_SoeRetroactivePayrollStatus)e.Status,
                Note = e.Note,
                StatusName = e.StatusName,
                NrOfEmployees = e.NrOfEmployees,
                TimePeriodName = e.TimePeriod?.Name ?? string.Empty,
                TimePeriodPaymentDate = e.TimePeriod?.PaymentDate,
                TimePeriodHeadId = e.TimePeriod?.TimePeriodHead?.TimePeriodHeadId ?? 0,
                TimePeriodHeadName = e.TimePeriod?.TimePeriodHead?.Name ?? string.Empty,
                Created = e.Created,
                CreatedBy = e.CreatedBy,
                Modified = e.Modified,
                ModifiedBy = e.ModifiedBy,
                State = (SoeEntityState)e.State,
            };

            if (e.RetroactivePayrollEmployee != null)
                dto.RetroactivePayrollEmployees = e.RetroactivePayrollEmployee.ToDTOs().ToList();
            if (e.RetroactivePayrollAccount != null)
                dto.RetroactivePayrollAccounts = e.RetroactivePayrollAccount.ToDTOs().ToList();

            return dto;
        }

        public static IEnumerable<RetroactivePayrollDTO> ToDTOs(this IEnumerable<RetroactivePayroll> l)
        {
            var dtos = new List<RetroactivePayrollDTO>();
            if (l != null)
            {
                foreach (var e in l)
                {
                    dtos.Add(e.ToDTO());
                }
            }
            return dtos;
        }

        public static TermGroup_SoeRetroactivePayrollStatus GetRetroactivePayrollStatus(this RetroactivePayroll e)
        {
            return e.RetroactivePayrollEmployee.ToList().GetRetroactivePayrollStatus();
        }

        #endregion

        #region RetroactivePayrollAccount

        public static RetroactivePayrollAccountDTO ToDTO(this RetroactivePayrollAccount e)
        {
            if (e == null)
                return null;

            #region Try load

            try
            {
                if (!e.IsAdded() && !e.AccountDimReference.IsLoaded)
                {
                    e.AccountDimReference.Load();
                    DataProjectLogCollector.LogLoadedEntityInExtension("RetroactivePayroll.cs e.AccountDimReference");
                }
            }
            catch (InvalidOperationException ex) { ex.ToString(); }

            #endregion        

            RetroactivePayrollAccountDTO dto = new RetroactivePayrollAccountDTO()
            {
                RetroactivePayrollAccountId = e.RetroactivePayrollAccountId,
                RetroactivePayrollId = e.RetroactivePayrollId,
                AccountDimId = e.AccountDimId,
                AccountId = null,
                Type = (TermGroup_RetroactivePayrollAccountType)e.Type,
                Created = e.Created,
                CreatedBy = e.CreatedBy,
                Modified = e.Modified,
                ModifiedBy = e.ModifiedBy,
                State = (SoeEntityState)e.State,
            };

            if (e.AccountDim != null)
                dto.AccountId = e.AccountDim.IsStandard ? e.AccountStdId : e.AccountInternalId;

            return dto;
        }

        public static IEnumerable<RetroactivePayrollAccountDTO> ToDTOs(this IEnumerable<RetroactivePayrollAccount> l)
        {
            var dtos = new List<RetroactivePayrollAccountDTO>();
            if (l != null)
            {
                foreach (var e in l)
                {
                    dtos.Add(e.ToDTO());
                }
            }
            return dtos;
        }

        #endregion

        #region RetroactivePayrollEmployee

        public static RetroactivePayrollEmployeeDTO ToDTO(this RetroactivePayrollEmployee e)
        {
            if (e == null)
                return null;

            RetroactivePayrollEmployeeDTO dto = new RetroactivePayrollEmployeeDTO()
            {
                RetroactivePayrollEmployeeId = e.RetroactivePayrollEmployeeId,
                RetroactivePayrollId = e.RetroactivePayrollId,
                ActorCompanyId = e.ActorCompanyId,
                EmployeeId = e.EmployeeId,
                EmployeeNr = e.Employee?.EmployeeNr ?? string.Empty,
                EmployeeName = e.Employee?.ContactPerson?.Name ?? string.Empty,
                Note = e.Note,
                Status = (TermGroup_SoeRetroactivePayrollEmployeeStatus)e.Status,
                Created = e.Created,
                CreatedBy = e.CreatedBy,
                Modified = e.Modified,
                ModifiedBy = e.ModifiedBy,
                State = (SoeEntityState)e.State,
            };

            if (e.RetroactivePayrollOutcome != null)
                dto.RetroactivePayrollOutcomes = e.RetroactivePayrollOutcome.ToDTOs().ToList();

            return dto;
        }

        public static IEnumerable<RetroactivePayrollEmployeeDTO> ToDTOs(this IEnumerable<RetroactivePayrollEmployee> l)
        {
            var dtos = new List<RetroactivePayrollEmployeeDTO>();
            if (l != null)
            {
                foreach (var e in l)
                {
                    dtos.Add(e.ToDTO());
                }
            }
            return dtos;
        }

        public static TermGroup_SoeRetroactivePayrollStatus GetRetroactivePayrollStatus(this List<RetroactivePayrollEmployee> l)
        {
            TermGroup_SoeRetroactivePayrollStatus retroPayrollStatus = TermGroup_SoeRetroactivePayrollStatus.Saved;

            List<RetroactivePayrollEmployee> retroEmployees = l.Where(i => i.State == (int)SoeEntityState.Active).ToList();
            int total = retroEmployees.Count;
            if (total > 0)
            {
                int nrOfStatusCalculated = retroEmployees.Count(i => i.Status == (int)TermGroup_SoeRetroactivePayrollEmployeeStatus.Calculated);
                int nrOfStatusPayroll = retroEmployees.Count(i => i.Status == (int)TermGroup_SoeRetroactivePayrollEmployeeStatus.Payroll);
                int nrOfStatusLocked = retroEmployees.Count(i => i.Status == (int)TermGroup_SoeRetroactivePayrollEmployeeStatus.Locked);

                bool statusLocked = nrOfStatusLocked > 0 && nrOfStatusLocked == total;
                bool statusPartlyLocked = nrOfStatusLocked > 0 && nrOfStatusLocked < total;
                bool statusPayroll = nrOfStatusPayroll > 0 && nrOfStatusPayroll == total;
                bool statusPartlyPayroll = nrOfStatusPayroll > 0 && nrOfStatusPayroll < total;
                bool statusCalculated = nrOfStatusCalculated > 0 && nrOfStatusCalculated == total;
                bool statusPartlyCalculated = nrOfStatusCalculated > 0 && nrOfStatusCalculated < total;

                if (Extensions.ExceedsThreshold(2, statusLocked, statusPartlyLocked, statusPayroll, statusPartlyPayroll, statusCalculated, statusPartlyCalculated))
                    retroPayrollStatus = TermGroup_SoeRetroactivePayrollStatus.Multiple;
                else if (statusLocked)
                    retroPayrollStatus = TermGroup_SoeRetroactivePayrollStatus.Locked;
                else if (statusPartlyLocked)
                    retroPayrollStatus = TermGroup_SoeRetroactivePayrollStatus.PartlyLocked;
                else if (statusPayroll)
                    retroPayrollStatus = TermGroup_SoeRetroactivePayrollStatus.Payroll;
                else if (statusPartlyPayroll)
                    retroPayrollStatus = TermGroup_SoeRetroactivePayrollStatus.PartyPayroll;
                else if (statusCalculated)
                    retroPayrollStatus = TermGroup_SoeRetroactivePayrollStatus.Calculated;
                else if (statusPartlyCalculated)
                    retroPayrollStatus = TermGroup_SoeRetroactivePayrollStatus.PartlyCalculated;
            }

            return retroPayrollStatus;
        }

        public static bool IsValidToCreateTransactions(this RetroactivePayrollEmployee rpe)
        {
            return (
                rpe.Status == (int)TermGroup_SoeRetroactivePayrollEmployeeStatus.Calculated &&
                rpe.State == (int)SoeEntityState.Active
                );
        }

        public static bool IsValidToDeleteTransactions(this RetroactivePayrollEmployee rpe)
        {
            return (
                rpe.RetroactivePayrollOutcome != null &&
                rpe.Status == (int)TermGroup_SoeRetroactivePayrollEmployeeStatus.Payroll &&
                rpe.State == (int)SoeEntityState.Active
                );
        }

        public static bool IsValidToDeleteOutcomes(this RetroactivePayrollEmployee rpe)
        {
            return (
                rpe.RetroactivePayrollOutcome != null &&
                rpe.State == (int)SoeEntityState.Active
                );
        }

        #endregion

        #region RetroactivePayrollOutcomeDTO

        public static RetroactivePayrollOutcomeDTO ToDTO(this RetroactivePayrollOutcome e)
        {
            if (e == null)
                return null;

            return new RetroactivePayrollOutcomeDTO()
            {
                RetroactivePayrollOutcomeId = e.RetroactivePayrollOutcomeId,
                RetroactivePayrolIEmployeeId = e.RetroactivePayrolIEmployeeId,
                ActorCompanyId = e.ActorCompanyId,
                EmployeeId = e.EmployeeId,
                ProductId = e.PayrollProductId,
                Quantity = e.Quantity,
                TransactionUnitPrice = e.TransactionUnitPrice,
                RetroUnitPrice = e.RetroUnitPrice,
                SpecifiedUnitPrice = e.SpecifiedUnitPrice,
                Amount = e.Amount,
                IsSpecifiedUnitPrice = e.IsSpecifiedUnitPrice,
                IsRetroCalculated = e.IsRetroCalculated,
                Created = e.Created,
                CreatedBy = e.CreatedBy,
                Modified = e.Modified,
                ModifiedBy = e.ModifiedBy,
                State = (SoeEntityState)e.State,
                ErrorCode = (TermGroup_SoeRetroactivePayrollOutcomeErrorCode)e.ErrorCode,
                ResultType = (TermGroup_PayrollResultType)e.ResultType,
                IsReversed = e.IsReversed,
                PayrollProductNumber = e.PayrollProduct?.Number ?? string.Empty,
                PayrollProductName = e.PayrollProduct?.Name ?? string.Empty,
            };
        }

        public static IEnumerable<RetroactivePayrollOutcomeDTO> ToDTOs(this IEnumerable<RetroactivePayrollOutcome> l)
        {
            var dtos = new List<RetroactivePayrollOutcomeDTO>();
            if (l != null)
            {
                foreach (var e in l)
                {
                    dtos.Add(e.ToDTO());
                }
            }
            return dtos;
        }

        public static bool IsValidForTransaction(this RetroactivePayrollOutcome e)
        {
            return (
                e.RetroUnitPrice.HasValue &&
                e.IsRetroCalculated &&
                e.State == (int)SoeEntityState.Active &&
                (e.ErrorCode == 0 || e.IsSpecifiedUnitPrice)
                );
        }

        #endregion
    }
}
